### PR TITLE
TensroIterator preserve format for binary, ternary operators.

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29461 [WIP] Experiments
* #29460 [WIP] Switch default of clone, to, *_like ops
* #29459 [WIP] Prepare copy op to switch memory format preservation defaults
* #28991 Adding function to convert Module to channels last
* #28544 Add torch.memory_format support to the TorchScript
* #28292 Add memory format support to the `resize_` op.
* **#28291 TensroIterator preserve format for binary, ternary operators.**
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* #27979 Add memory format support to `resize_as_` operator

Differential Revision: [D18044977](https://our.internmc.facebook.com/intern/diff/D18044977)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@ngimel I will add tests and benchmark results, so far I just looking your opinion on structure.

File: test/test_torch.py [hunk 1]
```diff
@@ -11956,6 +11957,88 @@ def test_memory_format_empty_like(self, device):
...
+            lambda x, y: x * y,
+            lambda x, y: y * x,
+            lambda x, y: x.sin(),
+            lambda x, y: x.sinh(),
+            lambda x, y: x.sqrt(),
+            # lambda x, y: abs(x), # https://github.com/pytorch/pytorch/issues/24531
+            # lambda x, y: x.abs(), # https://github.com/pytorch/pytorch/issues/24531
+            # lambda x, y: x.acos(), # https://github.com/pytorch/pytorch/issues/24532
+            lambda x, y: x.add(y, alpha=3),
+            lambda x, y: x.addcdiv(2, y, y),
```
Comment (ngimel) [inline test/test_torch.py]: given that y is contiguous, what are we testing here? Should it have been x, y in the arguments (same for addcmul)

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -275,33 +275,51 @@ DimVector TensorIterator::invert_perm(IntArrayRef input) const {
...
-      auto tensor_shape = invert_perm(shape_);
-      if (inverted) {
-        // can just return contiguous output
-        // it is faster because it avoids allocating 0 size tensor and resizing and restriding it
+      if (requires_channels_last_output_ && ndim() == 4) {
+        auto tensor_shape = invert_perm(shape_);
         op.tensor = at::empty(tensor_shape, op.options());
+        op.tensor.unsafeGetTensorImpl()->empty_tensor_restride(
+            MemoryFormat::ChannelsLast);
+        op.stride_bytes = apply_perm_and_mul(op.tensor.strides(), element_size);
```
Comment (ngimel) [inline aten/src/ATen/native/TensorIterator.cpp]: I cannot convince myself that `op.stride_bytes` and `op.tensor.strides()` are guaranteed to be consistent (one is defined by perm_ which can be whatever, and the other is hardcoded nhwc), and tests are only testing contiguity of output which is guaranteed by `empty_tensor_restride`, not its correctness.

File: test/test_torch.py (hunk 1 above)
Comment (VitalyFedyunin) [reply test/test_torch.py]: Addressed by adding more tests

File: aten/src/ATen/native/TensorIterator.cpp (hunk 2 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/TensorIterator.cpp]: Addressed by adding comment

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**